applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_multi_stage_kim_scheme.hpp

```diff
@@ -0,0 +1,989 @@
+        const TObjectType& rIntertiaVariable,
+        const VarComponentType& rFixVariable1,
+        const VarComponentType& rFixVariable2,
+        const VarComponentType& rFixVariable3,
+        const SizeType DomainSize = 3
+        )
+    {
+        Double3DArray& r_acceleration = itCurrentNode->FastGetSolutionStepValue(rAccelerationVariable);
+        Double3DArray& r_current_velocity = itCurrentNode->FastGetSolutionStepValue(rVelocityVariable);
+        const Double3DArray& r_current_residual = itCurrentNode->FastGetSolutionStepValue(rResidualVariable);
+
+        if (std::is_same<TObjectType,DoubleVarType>::value) {
+            const double& r_nodal_inertia = itCurrentNode->GetValue(NODAL_MASS);
+            const double& r_nodal_damping = itCurrentNode->GetValue(NODAL_DISPLACEMENT_DAMPING);
+
+            // Solution of the explicit equation:
+            if (r_nodal_inertia > numerical_limit)
+            {
+                noalias(r_acceleration) = (r_current_residual - r_nodal_damping * r_current_velocity) / r_nodal_inertia;
+            }
+            else
+                noalias(r_acceleration) = ZeroVector(3);
+        }
+        else if (std::is_same<TObjectType,ArrayVarType>::value) {
+            const Double3DArray& r_nodal_inertia = itCurrentNode->GetValue(NODAL_INERTIA);
+            const Double3DArray& r_nodal_rotational_damping = itCurrentNode->GetValue(NODAL_ROTATION_DAMPING);
+
+            const IndexType initial_k = DomainSize == 3 ? 0 : 2; // We do this because in 2D only the rotation Z is needed, then we start with 2, instead of 0
+            for (IndexType kk = initial_k; kk < 3; ++kk) {
+                if (r_nodal_inertia[kk] > numerical_limit)
+                    r_acceleration[kk] = (r_current_residual[kk] - r_nodal_rotational_damping[kk] * r_current_velocity[kk]) / r_nodal_inertia[kk];
+                else
+                    r_acceleration[kk] = 0.0;
+            }
+        }
+        else KRATOS_ERROR << "cannot handle input variable in \"explicit_multi_stage_kim_scheme.hpp\"" << std::endl;
+
+        std::array<bool, 3> fix_field = {false, false, false};
+        fix_field[0] = (itCurrentNode->GetDof(rFixVariable1, FieldPosition).IsFixed());
+        fix_field[1] = (itCurrentNode->GetDof(rFixVariable2, FieldPosition + 1).IsFixed());
+        if (DomainSize == 3)
+            fix_field[2] = (itCurrentNode->GetDof(rFixVariable3, FieldPosition + 2).IsFixed());
+
+        for (IndexType j = 0; j < DomainSize; j++) {
+            if (fix_field[j]) {
+                r_acceleration[j] = 0.0;
+            }
+        }
+    }
+
+    /**
+     * @brief This method updates the translation DoF
+     * @param itCurrentNode The iterator of the current node
+     * @param DisplacementPosition The position of the displacement dof on the database
+     * @param DomainSize The current dimention of the problem
+     */
+    void UpdateDegreesOfFreedomStage1(
+        NodeIterator itCurrentNode,
```

Comment:
usually I would recommend to pass a ref to the node unless you need the iterator

Reply:
I think we need the iterator here, just like we do in the central differencing